infrastructure/src/main/java/org/corfudb/infrastructure/management/HealingDetector.java

```diff
@@ -0,0 +1,235 @@
+     */
+    private String[] members;
+
+    /**
+     * Number of iterations to execute to detect a healed node in a round.
+     */
+    @Getter
+    @Setter
+    private int detectionThreshold = 3;
+
+    /**
+     * Threshold timeout to detect healing nodes.
+     * This should be kept lower if there is no tolerance for high latency recovering nodes.
+     */
+    @Getter
+    @Setter
+    private long detectionPeriodDuration = 5_000L;
+
+    /**
+     * Interval between iterations in a pollRound.
+     */
+    @Getter
+    @Setter
+    private long interIterationInterval = 1_000;
+
+    private int[] responses;
+    private IClientRouter[] membersRouters;
+    private CompletableFuture<Boolean>[] pollCompletableFutures = null;
+
+    private final long INVALID_EPOCH = -1L;
+    private final int INVALID_RESPONSE = -1;
+    private long currentEpoch = INVALID_EPOCH;
+
+    /**
+     * Executes the policy once.
+     * Checks for changes in the layout.
+     * Then polls all the servers and generates the report.
+     *
+     * @param layout Current Layout
+     */
+    public PollReport poll(Layout layout, CorfuRuntime corfuRuntime) {
+
+        // Performs setup and checks for changes in the layout to update unresponsive servers list.
+        setup(layout, corfuRuntime);
+
+        // Perform polling of all unresponsive servers.
+        return pollRound();
+
+    }
+
+    /**
+     * Triggered on every poll.
+     * Sets up the unresponsive servers array. (All servers in the unresponsiveServers list in
+     * the layout)
+     * Resets the currentEpoch to the layout epoch.
+     *
+     * @param layout       Latest known layout
+     * @param corfuRuntime Connected CorfuRuntime instance.
+     */
+    private void setup(Layout layout, CorfuRuntime corfuRuntime) {
+
+        currentEpoch = layout.getEpoch();
+
+        // Get all unresponsive servers.
+        members = layout.getUnresponsiveServers().stream().toArray(String[]::new);
+        Arrays.sort(members);
+
+        log.debug("Unresponsive members to poll, {}", new ArrayList<>(Arrays.asList(members)));
+        membersRouters = new IClientRouter[members.length];
+        responses = new int[members.length];
+
+        for (int i = 0; i < members.length; i++) {
+            try {
+                membersRouters[i] = corfuRuntime.getRouter(members[i]);
+                membersRouters[i].setTimeoutResponse(detectionPeriodDuration);
+            } catch (NetworkException ne) {
+                log.debug("Error creating router for {}", members[i]);
+            }
+        }
+        pollCompletableFutures = new CompletableFuture[members.length];
+    }
+
+    /**
+     * Each PollRound executes "detectionThreshold" number of iterations and generates the
+     * poll report.
+     *
+     * @return Poll Report with detected healed nodes.
+     */
+    private PollReport pollRound() {
+
+        Set<String> membersSet = new HashSet<>(Arrays.asList(members));
```

Comment:
Okay, I wrote some comments below, but after reading through this whole class, overall this class is way, _wayyyy_ to verbose for what it does.

You could express this entire class in a few lines, much more succinct, understandable and maintainable:

```java
public class HealingDetector implements IDetector {
   int numRounds = 3;
   Duration retryInterval = Duration.ofSeconds(10);

   @Override
   public PollReport poll(Layout layout, CorfuRuntime corfuRuntime) {
        Set<String> healedNodes = IntStream.range(0, 3)
                .mapToObj(i -> this.doRound(i, layout, runtime))
                .flatMap(Set::stream)
                .collect(Collectors.toSet());

        return new PollReport.PollReportBuilder()
                                             .pollEpoch(layout.getEpoch())
                                             .healedNodes(healedNodes)
                                             .build();
   }


    Set<String> doRound(int round, @Nonnull Layout layout, @Nonnull CorfuRuntime runtime) {
        Set<String> healed = layout.getUnresponsiveServers().stream()
            // Get pair of router strings -> ping success
            .map(s -> new AbstractMap.SimpleImmutableEntry<>(s, // key
                pingSuccessful(runtime.getRouter(s).getClient(BaseClient.class).ping()))) // value
            .filter(pair -> pair.getValue()) // only want successful pings
            .map(Map.Entry::getKey) // get router string
            .collect(Collectors.toSet());
       if (!healed.isEmpty()) { // Weird logic, but that's what the original code does.
           Sleep.sleepUninterruptibly(retryInterval);
       }
       return healed;
    }

    boolean pingSuccessful(@Nonnull CompletableFuture<Boolean> future) {
        try {
            return future.get();
        } catch (Exception e) {
            if (e.getCause() instanceof WrongEpochException) {
                return true;
            }
            return false;
        }
    }
}
```